Clean up public API

olafurpg opened the pull request:
- Remove unnecessary abstractions like `JsonRpcClient` or `JsonRpcNamedService`
- Rename LanguageServer into Server
- Rename LanguageClient into Client
- Add `final` where possible
- Add `private` where possible
- Remove `case` where possible
- Move as much as possible to `scala.meta.internal` as possible.

olafurpg committed:
Move to upickle for json.

The transitive dependency graph of upickle (including inlined jawn) is
~500kb while circe is ~8mb (primarily due to 22-tuple generated
implicits in cats). Considering that IntelliJ Scala team might want
to bring in lsp4s we should try to keep size of dependencies to a
minimum.

Quite a few changes were necessary to deliver this migration

- minimal "ping pong" unit test to verify the previous behavior
- `@json` macro annotation to replace `@JsonCodec`
- `Enums.findValues` def macro to replace enumeratum
- re-write all manual decoders/encoders by hand for upickle

While doing the refactoring, I took the liberty to minimize the lsp4s
public API

- Add `final` where possible
- Add `private` where possible
- Remove `case` where possible
- Move as much as possible to `scala.meta.internal` as possible.

olafurpg commented:
Still haven't manually tested this branch in metals repo, would like to finish that before merging this PR.

olafurpg committed:
Polish the API

This commit polishes the jsonrpc API by removing unnecessary vocabulary
and types and adding documentation to the most important methods and types.

- Use `scribe.LoggerSupport` instead of `scribe.Logger` to minimize risk
  of bincompat problems.
- s/LanguageServer/Server/
- s/LanguageClient/Client/
- s/JsonRpcClient/Client/
- Rename `JsonRpcNamedService` to `Service`, remove previous `Service` and `MethodName` traits
- Convert straight to `Array[Byte]` skipping intermediary `ujson.Js`
- Move private extension methods to `scala.meta.internal.jsonrpc._`
- More consistent names for `asJson*` extension methods for autocompletion-friendliness

olafurpg commented:
Working on changes to further refine the public API. Will try to keep the commit history clean, but I think it's best to do it all in one PR so that we don't spend time commenting on rough edges in this PR that are already fixed in my local commits.

gabro commented on macros/src/main/scala/scala/meta/jsonrpc/json.scala:
```diff
@@ -0,0 +1,56 @@
+        ) if c.mods.hasFlag(Flag.CASE) =>
+      q"""
+         $c
+         object $name extends { ..$early } with $parents { $self =>
```
Splice parents?

olafurpg replied:
Good catch, that was a bug 👍

olafurpg committed:
Splice parents

olafurpg commented:
For existing code that uses the metals jsonrpc and lsp modules, the main breaking changes are

```diff
- before
+ after

- import org.langmeta
+ import scala.meta
- new LanguageServer(...)
+ Server(...)
- new LanguageClient
+ Client(...)
- @JsonCodec case class Foo(...)
+ @json case class Foo(...)
```

olafurpg commented:
Testing the new API in metals repo with this WIP diff https://github.com/scalameta/metals/compare/master...olafurpg:lsp4s?expand=1 it compiles but the logs seems to be lost, will need more time to investigate.

ShaneDelmore commented on build.sbt:
```diff
@@ -5,17 +5,12 @@ inThisBuild(
-    organization := "org.scalameta",
-    licenses := Seq(
-      "Apache-2.0" -> url("http://www.apache.org/licenses/LICENSE-2.0")
-    ),
```
Why the license change?

olafurpg replied:
No license change, this line is duplicate of PublishPlugin

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/internal/jsonrpc/CancelParams.scala:
```diff
@@ -0,0 +1,6 @@
+import scala.meta.jsonrpc.json
+import ujson.Js
+
+@json case class CancelParams(id: Js)
```
What is the decision making process for when to use case class or not now?  Are they ok as long as they are internal?

olafurpg replied:
Avoid case class when possible, unless it's internal API in which case it's fine.

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/internal/jsonrpc/package.scala:
```diff
@@ -0,0 +1,41 @@
+      catch { case NonFatal(ex) => Left(ex) }
+
+    /** Converts T into ujson.Js with ReadWriter[T] writer */
+    def asJsonEncoded(implicit writer: ReadWriter[T]): Js = writeJs(e)
```
AsJsonEncoded makes me think this is a string representing json.  Since it’s actually a Js type how about asJs or asJson?

olafurpg replied:
`asJson` would be ambiguous from the `asJsonParsed`, I agree the name is a bit confusing but I optimized for autocompletion in the IDE 😅 open for suggestions

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/internal/jsonrpc/package.scala:
```diff
@@ -0,0 +1,84 @@
+  implicit class XtensionAsSyntax(js: Js) {
+
+    /** Converts ujson.Js to T with ReadWriter[T] reader */
+    def asJsonDecoded[T](implicit reader: ReadWriter[T]): Either[Throwable, T] =
```
Why Either[Throwable, T] instead of Try[T]?

olafurpg replied:
Because previously we had either from circe and this didn't break for comprehensions

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/jsonrpc/Endpoint.scala:
```diff
@@ -1,30 +1,77 @@
+/**
+ * A single JSON-RPC method with it's request and response types.
+ *
+ * A protocol like LSP is in essence a collection of endpoints.
```
This text confuses me but I don’t have constructive feedback on how to improve it.  It may be due to my background, but when you say a collection of endpoints, I immediately think of routes, thinking each endpoint is a different route where with lsp my intuition is that there is one endpoint which receives different messages and dispatches on them.  It feels very OO in that way, where I have a single “receive” endpoint which dispatches based on the message received.

The functionality doesn’t confuse me though, just the terminology.  I guess the endpoint is what you are routing/dispatching to.

olafurpg replied:
LSP is kind of like `List[Endpoint]`, and you can treat each endpoint as an independent routes. That's at least how it feels like using this library

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/jsonrpc/InputOutput.scala:
```diff
@@ -0,0 +1,10 @@
+/** Wrapper around a pair of input/output streams. */
+final class InputOutput(val in: InputStream, val out: OutputStream) {
+  def withIn(in: InputStream): InputOutput = new InputOutput(in, out)
+  def withOut(out: OutputStream): InputOutput = new InputOutput(in, out)
```
I know Sebastien advocates this pattern but I fail to see any value in it here.  That said, I don’t see much expense to it either, but it definitely feels like speculative code that we don’t actually have a use for, but think there may be a use for later.

olafurpg replied:
Removed these helpers, I agree they're not necessary. If the constructor was private or I anticipated new fields I would keep these methods

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/jsonrpc/Message.scala:
```diff
@@ -1,72 +1,74 @@
 }
 
-@JsonCodec case class Request(
+@json final case class Request(
```
This is not internal or private, but is a case class not abstract class, can you explain when to use each?

olafurpg replied:
- case classes are fine if you expect they will never add new fields
- case classes should always be final
- don't use case classes if you expect to add new fields

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/jsonrpc/Message.scala:
```diff
@@ -1,72 +1,74 @@
+/**
+ * Supertype for all response types.
+ *
+ * @note Effectively sealed but not sealed because upickle automatically includes `$type` field for subtypes.
```
Is there no way to configure upickle to not add the $type field?

ShaneDelmore commented on jsonrpc/src/main/scala/scala/meta/jsonrpc/Services.scala:
```diff
@@ -1,120 +1,59 @@
+      new Function[A, Task[Either[Response.Error, B]]] {
+        override def apply(v1: A): Task[Either[Response.Error, B]] =
+          Task(Right(f(v1)))
+      }
```
Could you. Use type ascription to replace the new Function?
`a: A => Task(Right(f(a)): Either[Response.Error, B])`

olafurpg replied:
Removed `new Function`, I had it because I was refactoring from `new Service`, a lambda is much cleaner.

olafurpg committed:
Un-private useful methods

olafurpg committed:
Move LanguageServer alias to correct package.

olafurpg committed:
Use Double instead of Long

JSON number does not fit longs so upickle encodes them as strings.

olafurpg committed:
Fix issue where upickle tags json with `$type` key.

We don't use automagic serialization for sealed ADT so this commits
disabled tagging writers/readers with subtypes of sealed traits.

olafurpg committed:
Write to ujson.Js instead of String.

This is a nasty bug and I'm unhappy this compiled before.  Hover showed
the json string value of the JSON-PRC "result" in the response.

ShaneDelmore commented on macros/src/test/scala/tests/jsonrpc/JsonSuite.scala:
```diff
@@ -0,0 +1,47 @@
+    camelCase: String = ""
+)
+object User extends Foo with Bar {
+  println(bar)
```
Is this println somehow part of the test suite or just a leftover?

ShaneDelmore reviewed (approved):
Functionality looks good as usual.   I had a few suggestions and questions on naming and style, no blockers.

olafurpg committed:
Address review

olafurpg reviewed (commented):
Thanks for the review @ShaneDelmore !

I recently pushed a fix for the `$type` issue, I'm manually testing the change and most things seem to work fine but still requires a bit more polish.

I am starting to get cold feet about changing json library. upickle has a lot nice going for it but I personally enjoyed using circe more

- better decoding error messages
- less surprising behavior like implicit conversions from `String` to `Js.Str` causing subtle runtime bugs when mixing `write` with `writeJs`.

Considering that monix v3 requires cats either way, I feel like we can't avoid it eventually. To minimize the classpath, it's probably possible to tree shake out all the unused implicits boilerplate bringing it down to ~1-2mb. Any thoughts @jastice?

ShaneDelmore commented:
👍 to keeping circe and just dropping the unused implicits if it doesn’t end up being too much effort.

jastice commented:
Keeping circe and cats sounds reasonable to me, especially if it can be taken to < 5MB and it makes upgrading Monix easier

olafurpg committed:
Revert back to circe

upickle has a lot nice going for it but I personally enjoyed using
circe more

- better decoding error messages
- less surprising behavior like implicit conversions from String
to Js.Str causing subtle runtime bugs when mixing write with
writeJs.

Given that monix v3 requires cats either way it seems like an inevitable
dependency.

olafurpg commented:
Sounds good, let's keep circe in that case. At least we looked into alternatives

olafurpg committed:
Revert back to enumeratum

olafurpg committed:
Add test suite for header parsing.

olafurpg commented:
This PR no longer changes dependencies, it only polishes the public API with documentation and hopefully improved names and removed redundant APIs.

Any ideas why sbt is hanging indefinitely after "Done compiling"? https://travis-ci.org/scalameta/lsp4s/jobs/373219479#L531 only happens in 2.12, not 2.11

olafurpg commented:
Nevermind this PR, given it's not minimizing dependencies I no longer see value in the refactoring.
